### ISSUE TITLE
Alternative safer approach to composing disjunctions (OR queries)

### ISSUE THREAD

**Issue by moll** [tgriesser/knex#1714]
Hey,

I'm not sure if there's an existing topic on the design of the query interface, but if there is, please direct me there. I must've missed it. Also, rather than email Tim, @tgriesser, directly I thought I'd create an issue here for everyone to comment on. I hear there's v1 on the way.

I propose a change to the way disjunction queries — `OR` — are built. The way they're constructed now, with `orWhere`, I'd argue, is way too dangerous. The biggest benefit of a query interface, as opposed to typing in SQL, is composability: it allows different functions to build up parts of a query that you'll later combine. Functions, for the most part, need not know of each other's expressions.

However, a single `orWhere` wrecks the whole show. Imagine initializing a query with the user account:

``` js
var query = knex.where("account_id", 3)
```

Then passing that on to another function for further filters. 

``` js
query = query.where("name", "John").orWhere("email", "john@example.com")
```

Should that function innocently use `orWhere`, you've accidentally introduced a security issue: The resulting SQL from above creates a conjunction of `account_id = 3 and name = 'John'` and `OR`s that with `email = 'john@example.com'`. The syntax for what we really wanted would've been to pass a function to `where` and do the disjunction inside that. Basically, one should _always_ use the function-form of `where` for safety.

Now, I argue, the existence of `orWhere` is so **error prone and risky** that I suggest redesigning the `orWhere` approach and replacing it with something that never creates unexpected disjunctions. In pseudo code, something like:

``` js
var knex = Knex(...)
var query = knex.where("account_id", 3)
query = query.any([knex.where("name", "John"), knex.where("email", "john@example.com")])
query.from("models").then(...)
```

For the above to work, Knex would have to be able to merge statements from two queries, but I think that's easy enough to do. Sure, some might pass unsupported queries (e.g. with joins) to the imaginary `QueryBuilder.prototype.any`, but those would be easy to throw exceptions on.

I wrote a simple [immutable] query object some years ago with a similar API to wrap Knex's and it's been very convenient and safe to use. However, I don't really want to be in the business of having my own, so I thought I'd perhaps advocate making Knex safer and more composable. ;)

Cheers

**Comment by absolux** [comment]
Hey @moll 
`any` is a keyword in SQL, and adding a method like that can be comfusing, in my opinion.
So, I suggeste you to use functions instead like
```js
query.orWhere(qb => {
  qb.where("name", "John").orWhere("email", "john@example.com")
})
```

**Comment by LachlanStuart** [comment]
I'm know I'm late to this conversation, but conventional security wisdom is to always try to get framework-level solutions first, such as the safer disjunction construct that @moll proposed. If framework-level solutions aren't available, then one should try to have clear security boundaries within the application code so that code that is potentially insecure stands out.

You can never trust your coworkers or the future developers of your codebase to follow complex rules like "`orWhere` clauses must always be wrapped with a `where` callback", as they're hard to validate and aren't obvious even when reviewing other peoples' code. It's a bit more plausible to have a rule like "sensitive tables must be accessed through their respective APIs", as direct access to certain tables will start to stand out as potential security issues.

The best compromise I've found with knex is to wrap filters for sensitive tables in a subquery callback, e.g.:
```js
function pgItemsVisibleToUser(user) {
  return subquery => {
    let filter;
    if (user != null && user.role === 'admin') {
      filter = item => true;
    } else if (user != null) {
      filter = item => item.where('is_public', true)
                            .orWhere('author', user.email);
    } else {
      filter = item => ds.where('is_public', true);
    }
    return subquery.from('item').where(filter).as('item');
  };
}
```

Which can then be used in queries without risk of the `orWhere` clauses breaking the filter, e.g. 
```js
knex
    .from('item_tag')
    .join(pgItemsVisibleToUser(user), 'item_tag.item_id', 'item.item_id')
    .where('item_tag.tag', 'book')
    .orWhere('item.type', 'literature');
```
to produce SQL like:
```sql
select * 
from "item_tag" 
inner join (select * from "item" where ("is_public" = true or "author" = 'foo@bar.baz')) as "item" 
    on "item_tag"."item_id" = "item"."item_id" 
where "item_tag"."tag" = 'book' 
    or "item"."type" = 'literature'
```

**Comment by elhigu** [comment]
I think that the current `or` -API is enough for such low level library that knex is. It reflects closely to the generated SQL. If knex someday suports extending its APIs, then this may be implemented as query builder extension.

**Comment by moll** [comment]
Sorry, @elhigu, but I think we must not naively dismiss of what I argue is a very poor API from a security perspective like this. SQL injections and query vulnerabilities of all sorts form _the_ most common security issue in webapps. Query builders _exist_ to make composition easy and safe. We've got easy. We're not there with safety yet. Surely you wouldn't claim vulnerabilities through value interpolations are "fine" in a "low level library". Why would you claim they're fine in equally risky query construction?

I'm glad we agree on solving security issues in the API level, @LachlanStuart. I do think your distinction of sensitive tables from insensitive ones is not a helpful, though. First bugs are bugs regardless of whether they involve people's private data or not. Secondly, following the requirement of wrapping all access to tables in functions merely pushes the problem down _and_ entirely defeats the purpose of query builders --- convenient composition.

@tgriesser, what's your view on security and safe APIs?

**Comment by elhigu** [comment]
@moll No, problem. I still disagree, that knex core should have yet another query API like this. Knex's inconsistent APIs are already a problem and this kind of designing/tuning some minor detail a bit by bit further is not helping at all for providing consistent and well defined API for your query building. 

If there would be bigger more complete carefully thought "future query API rehaul / re-design" with deprecation of old APIs then I could be pro this. I don't see how this would help the current usage, because people who doesn't know that they need to use parenthesis correctly would keep on writing: 

```js
query = query.orWhere("name", "John").orWhere("email", "john@example.com")
```

instead of

```js
query = query.where(builder => builder.where("name", "John").orWhere("email", "john@example.com"))
```

or 

```js
query = query.any([knex.where("name", "John"), knex.where("email", "john@example.com")])
```

Anyways this is now out of my hands so other collaborators can choose what to do with this.

**Comment by moll** [comment]
> If there would be bigger more complete carefully thought "future query API rehaul / re-design" with deprecation of old APIs then I could be pro this. I don't see how this would help the current usage, because people who doesn't know that they need to use parenthesis correctly would keep on writing.

Oh, I agree with you there. Leaving risky APIs lingering around would defeat the purpose. Much like the unbounded string copy routines in C that everyone _should_ know not to use but occasionally still do.

**Comment by mercmobily** [comment]
The current `orWhere` API can lead to unintended query vulnerabilities when developers forget to properly group their conditions.
The issue highlighted is indeed serious. When composing queries dynamically, forgetting to wrap OR conditions in parentheses can completely bypass security filters:

```javascript
// Vulnerable code - OR breaks the account_id filter
let query = knex("items").where("account_id", 3);
if (filters.name) query = query.orWhere("name", filters.name);
// Results in: WHERE account_id = 3 OR name = 'foo'
// This returns items from ALL accounts where name matches!
```

**Current workaround**

The current safe approach requires disciplined use of callbacks:
```javascript
let query = knex("items").where("account_id", 3);
if (filters.name || filters.email) {
  query = query.where(qb => {
    if (filters.name) qb.where("name", filters.name);
    if (filters.email) qb.orWhere("email", filters.email);
  });
}
// Results in: WHERE account_id = 3 AND (name = 'foo' OR email = 'bar')
```

**Your proposed solution**

Your `any()` and `all()` methods would make the safe pattern more explicit and harder to misuse:
```javascript
query = query.any([
  knex.where("name", filters.name),
  knex.where("email", filters.email)
]);
```

**The challenge**

As @elhigu points out, adding new APIs whilst keeping the risky ones would only create more confusion. Developers unaware of the security implications would continue using `orWhere` directly. A proper solution would require:

1. Deprecating direct `orWhere` usage (breaking change)
2. Providing migration guides and tooling
3. Ensuring the new API covers all use cases

This would be a significant breaking change affecting virtually every Knex application. Whilst the security benefits are clear, the migration cost would be enormous.

**Alternative approaches**

Rather than changing the core API, we could:

1. Add prominent security warnings to the documentation
2. Create a linting rule to detect potentially unsafe `orWhere` usage
3. Provide a query builder extension for security-critical applications
4. Add helper methods that make the safe pattern easier without removing the existing API

I would start with (1) and (2), and then make new tickets for (3) and (4) once (1) and (2) are done
